File: cpp/279.cpp
#include<iostream>
using namespace std;

class Solution {
public:
  int numSquares(int n) {
    int cnt[n+1];
    for(int i=0; i<=n; i++)
      cnt[i] = i;
    for(int i=0; i<=n; i++){
      for(int j=1; j*j<=i; j++){
        int sq = j*j;
        if(cnt[i-sq]+1 < cnt[i])
          cnt[i] = cnt[i-sq]+1;
      }
    }
    return cnt[n];
  }
};

int main(){
  Solution sol = Solution();
  cout << sol.numSquares(7927) << endl;
}

File: python/279.py
class Solution(object):
    cache = []
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        cnt = self.cache
        ln = len(cnt)
        cnt.extend(range(ln, n+1))
        for i in xrange(ln, n+1):
            for j in range(1, i+1):
                sq = j*j
                if sq > i:
                    break
                if cnt[i-sq] + 1 < cnt[i]:
                    cnt[i] = cnt[i-sq] + 1
        return cnt[n]

sol = Solution()
print sol.numSquares(7927)

